### PR TITLE
WIP, ENH: First draft of nansum using _nan_mask()

### PR DESCRIPTION
**Pull request by tylerjereddy**
Starting on the third step of #12662.

Working on this has revealed an issue with the drafted private function with respect to 0 D arrays. I seem to have patched that for now.

Running the full test suite seems to reveal another little treat -- `np.matrix` doesn't seem to play well with my initial changes to nansum.

Probably a bit early for benchmarks, but from a quick local check:
`asv continuous -e -b "Nan" master nansum_where >& benchresults.txt`

Produces:

```
       before           after         ratio
     [b5b47a39]       [d83b05a0]
+       174±0.9μs        211±0.8μs     1.22  bench_lib.Nan.time_nansum(200000, 0.1)
+         171±1μs          206±2μs     1.20  bench_lib.Nan.time_nansum(200000, 0)
-     6.82±0.04μs      6.18±0.05μs     0.91  bench_lib.Nan.time_nansum(200, 0)
-      6.83±0.1μs      6.12±0.04μs     0.90  bench_lib.Nan.time_nansum(200, 2.0)
-     6.76±0.08μs      6.05±0.02μs     0.89  bench_lib.Nan.time_nansum(200, 0.1)
-     7.30±0.07μs      6.40±0.05μs     0.88  bench_lib.Nan.time_nansum(200, 50.0)
-     7.12±0.06μs      6.21±0.08μs     0.87  bench_lib.Nan.time_nansum(200, 90.0)
-        564±10μs         437±20μs     0.78  bench_lib.Nan.time_nansum(200000, 90.0)
```

### CODE REVIEW

**Comment by tylerjereddy** [comment]
I adjusted to avoid `np.matrix` with the new reduction behavior and re-ran the benchmarks:

`asv continuous -e -b "Nan" master nansum_where >& benchresults.txt`

```
       before           after         ratio
     [2f2dfa19]       [46bf17ca]
     <master>         <nansum_where>
+         164±2μs          209±6μs     1.28  bench_lib.Nan.time_nansum(200000, 0.1)
-      8.42±0.1μs      7.41±0.06μs     0.88  bench_lib.Nan.time_nansum(200, 2.0)
-      8.40±0.3μs      7.30±0.06μs     0.87  bench_lib.Nan.time_nansum(200, 0.1)
-        592±10μs         442±10μs     0.75  bench_lib.Nan.time_nansum(200000, 90.0)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

So, some things get slightly faster, and a large array with small % np.nan gets slower. Not sure if that is what @mhvk would expect.

**Comment by tylerjereddy** [comment]
I'll close this PR of mine -- there's not much evidence of remarkable performance improvement, and many of the other nan functions would have more complicated implementation details that are unlikely to make the relative performance much better. It doesn't seem to be a priority either.

**Comment by mhvk** [comment]
Hmm, I actually really liked this PR - just having had a bit of a hectic period so less change to review.  Let me re-open it for now, since I think there is a larger discussion to be had: currently, the `nanfunctions` cannot be used easily by array-likes unless they overwrite every numpy function using `__array_function__`. With your PR, things will work much more easily, as it will just rely on `__array_ufunc__`.

Also: on the speed, with this PR, `nanfunctions` will benefit always from any improvement made in the ufuncs more generally.

File: numpy/lib/nanfunctions.py [hunk 1]
```diff
@@ -640,8 +641,15 @@ def nansum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
     nan
 
     """
-    a, mask = _replace_nan(a, 0)
-    return np.sum(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims)
+    try:
+        a = np.asanyarray(a)
```
Comment (mhvk) [inline numpy/lib/nanfunctions.py]: Is this `asanyarray` necessary for any test? I would have thought `np.sum` would take care.

File: numpy/lib/nanfunctions.py [hunk 2]
```diff
@@ -640,8 +641,15 @@ def nansum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
...
+        return np.sum(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims,
+                      where=_nan_mask(a))
+    except TypeError:
+        # retain old mask handling for np.matrix,
```
Comment (mhvk) [inline numpy/lib/nanfunctions.py]: It seems strange to add work-arounds for code that we are deprecating. I think at least a deprecation-warning might be reasonable. I'll ping the mailing list.

**Review by mhvk — COMMENTED** [review]
It seems like there is a larger question here that we need to answer...

File: numpy/lib/nanfunctions.py (hunk 1 above)
Comment (shoyer) [reply numpy/lib/nanfunctions.py]: As I noted on the mailing list, this should probably be `np.asarray(a)` instead, which would obviate the need for the fallback logic.

Comment (mhvk) [reply numpy/lib/nanfunctions.py]: Yes, that would avoid the `matrix` problems and keep the behaviour for duck arrays the same, but break other subclasses such as `MaskedArray` and `Quantity` (which work fine with the present implementation), and will not allow duck arrays to override by updating their `sum` method or correctly implementing `__array_ufunc__`.

But what would work is to use your suggestion for the fall-back logic:
```
try:
    return np.sum(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims,
                  where=_nan_mask(a))
except TypeError:
    return np.asarray(a).sum(....)
```

Comment (mhvk) [reply numpy/lib/nanfunctions.py]: Actually, as noted in the thread, the `asarray` call is sadly not OK as we currently use `asanyarray` - and it thus won't solve the fact that `matrix` cannot handle it.

**Comment by mhvk** [comment]
After a long discussion at https://mail.python.org/pipermail/numpy-discussion/2019-June/079633.html, the conclusion would seem to be that the benefits of cleaning up `nanfunctions` are not worth the complications of dealing with deprecations.

**Pull request closed without merging**